File: src/contests/cg/puzzle/go/telephone-numbers.go
// 🎮 CodinGame Puzzle - telephone-numbers
// https://www.codingame.com/training/medium/telephone-numbers

package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// Node is a trie node: a map from digit to child node.
type Node map[byte]Node

func main() {
	reader := bufio.NewReader(os.Stdin)
	var n int
	fmt.Fscan(reader, &n)
	reader.ReadString('\n') // consume rest of the first line
	root := Node{}
	cable := 0
	for t := 0; t < n; t++ {
		line, _ := reader.ReadString('\n')
		number := strings.TrimRight(line, "\r\n")
		node := root
		for i := 0; i < len(number); i++ {
			d := number[i]
			if _, ok := node[d]; !ok {
				node[d] = Node{}
				cable++
			}
			node = node[d]
		}
	}
	fmt.Println(cable)
}

File: src/contests/cg/puzzle/dart/telephone-numbers.dart
// 🎮 CodinGame Puzzle - telephone-numbers
// https://www.codingame.com/training/medium/telephone-numbers

import 'dart:io';

void main() {
  int n = int.parse(stdin.readLineSync()!.trim());
  Map<String, dynamic> root = {};
  int cable = 0;
  for (int i = 0; i < n; i++) {
    String number = stdin.readLineSync()!.trim();
    Map<String, dynamic> node = root;
    for (int j = 0; j < number.length; j++) {
      String d = number[j];
      if (!node.containsKey(d)) {
        node[d] = <String, dynamic>{};
        cable++;
      }
      node = node[d] as Map<String, dynamic>;
    }
  }
  print(cable);
}
